src/ls_odbc.c

```diff
@@ -601,32 +600,135 @@ static int create_connection (lua_State *L, int o, env_data *env, SQLHDBC hdbc)
 	return 1;
 }
 
+/*
+** Uses a DSN string to connect to a ODBC source dynamically
+** Lua Input: {
+**    dsn = <DSN string>,
+** }
+** Lua Returns:
+**   connection object if successfull
+**   nil and error message otherwise.
+*/
+static int env_table_connect_DSN (lua_State *L)
+{
+	env_data *env = getenvironment (L);
+	SQLCHAR *sourcename = (SQLCHAR *)luasql_table_optstring(L, 2, "dsn", NULL);
+
+	SQLHDBC hdbc;
+	SQLCHAR sqlOutBuf[4097];
+	SQLSMALLINT sqlOutLen;
+	SQLRETURN ret;
+
+	ret = SQLSetEnvAttr (env->henv, SQL_ATTR_ODBC_VERSION, (void *)SQL_OV_ODBC3, 0);
+	if (error(ret)) {
+		return luasql_faildirect (L, "error setting SQL version.");
+	}
+
+	/* tries to allocate connection handle */
+	ret = SQLAllocHandle (hDBC, env->henv, &hdbc);
+	if (error(ret)) {
+		return luasql_faildirect (L, "connection allocation error.");
+	}
+
+	/* tries to connect handle */
+	ret = SQLDriverConnect (hdbc, NULL, sourcename, SQL_NTS, sqlOutBuf, 4096,
+	                        &sqlOutLen, SQL_DRIVER_NOPROMPT);
+	if (error(ret)) {
+		ret = fail(L, hDBC, hdbc);
+		SQLFreeHandle(hDBC, hdbc);
+		return ret;
+	}
+
+	/* success, return connection object */
+	ret = create_connection (L, 1, env, hdbc);
+	if(ret == 1) {
+		/* Add the sqlOutBuf string to the results, for diagnostics */
+		lua_pushlstring(L, (char *)sqlOutBuf, sqlOutLen);
+		return 2;
+	}
+
+	return ret;
+}
 
 /*
-** Creates and returns a connection object
-** Lua Input: source [, user [, pass]]
+** Reforms old connection style to new one
+** Lua Input: source, [user, [pass]]
 **   source: data source
 **   user, pass: data source authentication information
 ** Lua Returns:
+**   new connection details table
+*/
+static void env_connect_fix_old (lua_State *L)
```

Comment:
Suggestion: replace terms such as "old" and "new"; over the years these definitions get outdated and stop making sense (e.g. what if a third way of connecting comes up? becomes `newer`?). I suggest `env_connect_convert_args_to_table` instead.

Also, "fix" is a heavy term, because it implies something was "broken", when in this case it was clearly not.